include/faabric/scheduler/DistributedCoordination.h

```diff
@@ -0,0 +1,81 @@
+#pragma once
```

Comment:
My biggest design concern re. the `DistributedCoordination` class is that it is shared among all applications. Thus forcing us to implement a distributed coordination to access the `DistributedCoordination` which feels a bit unnatural.

I am wondering, could we have a per-app instance of the `DistributedCoordination` class? I imagine there must be some strong argument against this, but just pointing it out. 

Plus, I am assuming that contention will be a problem, but I don't have empirical proof which may be a case of pre-optimisation.

Reply:
Yes, all good points, although I think coordinating access to the coordinator will always be required, even if it's one per app, as there will be multiple executors from that app on any given host.

Whether contention is a bottleneck will depend on the workload. If it's something CPU-intensive which performs distributed coordination at the end of a big calculation to share results, it won't be a problem. If it's something with an atomic section nested deep in some inner loop that's called thousands of times, then it will be a problem. 

A bigger design flaw that I've spotted is that we actually need multiple coordinators _per app_, to handle nested multi-threaded constructs. As a result, I'll need to change it to look up coordinators on an arbitrary ID, rather than the app ID. I'll re-request a review and update the description once this is done.